demos/mantle_convection/adaptive_base_case/adaptive_base_case.py

```diff
@@ -0,0 +1,362 @@
+
+X = SpatialCoordinate(mesh)
+T_init = 1.0 - X[1] + 0.05 * cos(pi * X[0]) * sin(pi * X[1])
+u_init = as_vector((0., 0.))
+p_init = 0.
+
+# In all other demos, we would now continue defining some function spaces,
+# create some functions, and then the solvers using those. As we will be
+# adapting the mesh these steps will need to be repeated after each adapt.
+# Therefore we will wrap these steps in a python function. The following python
+# function in fact wraps the entire base case: it takes in a mesh builds up the
+# function spaces, functions and solvers and then runs the model for a
+# specified number of timesteps. We will be using this function in an outside
+# loop, where we adapt the mesh and then call the function to run the model on
+# the adapted mesh for a number of timesteps. The temperature, velocity
+# and pressure solution at the end of those timesteps, which we have solved
+# on the current adapted mesh, will need to be interpolated after the next mesh
+# adaptivity stage onto the new mesh and are therefore provided as `T_init`,
+# `u_init`, and `p_init` arguments. In the first call to this subroutine, the mesh
+# is still the initial mesh, and we simply use the initial condition expressions
+# we have just defined.
+
+# +
+
+
+def run_interval(mesh, time, timestep, Nt, T_init, u_init, p_init):
+    """Run for Nt timesteps on the given mesh
+
+    This sets up all the usual functionspace, equations, solvers, etc.
+    on the given mesh, and runs the timeloop for the given n/o timesteps.
+    Everything is exactly like in the base case.
+
+    args:
+    mesh - mesh to solve the equations on
+    time - starting time (for logging purposes)
+    timestep - starting timestep no. (for logging purposes)
+    Nt - n/o timesteps to run
+    T_init, u_init, p_init - initial conditions, or last solution from previous mesh
+                             we interpolate these onto the current mesh
+
+    returns:
+    time, T, u, p - time, temperature, velocity and pressure at the end of the timesteps
+    steady_state_reached - have we reached steady state during the timesteps?
+    """
+
+    # Set up function spaces - currently using the bilinear Q2Q1 element pair:
+    V = VectorFunctionSpace(mesh, "CG", 2)  # Velocity function space (vector)
+    W = FunctionSpace(mesh, "CG", 1)  # Pressure function space (scalar)
+    Q = FunctionSpace(mesh, "CG", 1)  # Temperature function space (scalar)
+    Z = MixedFunctionSpace([V, W])  # Mixed function space.
+
+    z = Function(Z, name='Solution')  # A field over the mixed function space Z.
+
+    # Output function space information:
+    log("Number of Velocity DOF:", V.dim())
+    log("Number of Pressure DOF:", W.dim())
+    log("Number of Velocity and Pressure DOF:", V.dim()+W.dim())
+    log("Number of Temperature DOF:", Q.dim())
+
+    # Set up temperature field and extract velocity and pressure from "mixed" function z
+    T = Function(Q, name="Temperature")
+    u, p = z.subfunctions
```

Comment:
We need to explain here why it is OK to use `subfunctions` instead of `split`. Ideally, we could have both and highlight where the subfunction is needed.

Reply:
I think I tried with `split` and ran into issues. If I remember well, the interpolation steps fail. So I am happy to keep `subfunctions`, but we should highlight why we take a different approach here.